File: Component/BusiLib/src/main/java-voice/com/component/voice/control/VoiceControlPanelView.java
package com.component.voice.control;

import android.content.Context;
import android.content.res.TypedArray;
import android.support.v7.widget.AppCompatRadioButton;
import android.util.AttributeSet;
import android.widget.RadioGroup;
import android.widget.ScrollView;
import android.widget.SeekBar;

import com.common.log.MyLog;
import com.common.utils.U;
import com.common.view.ex.ExTextView;
import com.component.busilib.R;
import com.engine.EngineEvent;
import com.engine.Params;
import com.zq.mediaengine.kit.ZqEngineKit;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

public class VoiceControlPanelView extends ScrollView {
    public final String TAG = "VoiceControlPanelView";

    ExTextView mPeopleVoice;
    SeekBar mPeopleVoiceSeekbar;
    ExTextView mAccVoice;
    SeekBar mMusicVoiceSeekbar;

    RadioGroup mScenesBtnGroup;
    AppCompatRadioButton mDefaultSbtn;
    AppCompatRadioButton mKtvSbtn;
    AppCompatRadioButton mRockSbtn;
    AppCompatRadioButton mDianyinSbtn;
    AppCompatRadioButton mKonglingSbtn;

    // 记录值用来标记是否改变
    Params.AudioEffect mBeforeMode;
    int mBeforePeopleVoice;
    int mBeforeMusicVoice;

    Params.AudioEffect mAfterMode;
    int mAfterPeopleVoice;
    int mAfterMusicVoice;

    boolean isShowACC = true;  //是否显示伴奏

    public VoiceControlPanelView(Context context) {
        super(context);
        init(context, null);
    }

    public VoiceControlPanelView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context, attrs);
    }

    protected int getLayout(){
        return R.layout.voice_control_panel_layout;
    }

    protected int getMarginLeft(){
        return U.getDisplayUtils().getScreenWidth() - U.getDisplayUtils().dip2px(30 + 24) - U.getDisplayUtils().dip2px(53 * 5);
    }

    public void init(Context context, AttributeSet attrs) {
        TypedArray typedArray = context.obtainStyledAttributes(attrs, R.styleable.VoiceControlPanelView);
        isShowACC = typedArray.getBoolean(R.styleable.VoiceControlPanelView_isShowACC, true);
        typedArray.recycle();

        inflate(getContext(), getLayout(), this);

        mPeopleVoice = (ExTextView) this.findViewById(R.id.people_voice);
        mPeopleVoiceSeekbar = (SeekBar) this.findViewById(R.id.people_voice_seekbar);
        mAccVoice = (ExTextView) this.findViewById(R.id.acc_voice);
        mMusicVoiceSeekbar = (SeekBar) this.findViewById(R.id.music_voice_seekbar);

        mDefaultSbtn = (AppCompatRadioButton) this.findViewById(R.id.default_sbtn);
        mKtvSbtn = (AppCompatRadioButton) this.findViewById(R.id.ktv_sbtn);
        mRockSbtn = (AppCompatRadioButton) this.findViewById(R.id.rock_sbtn);
        mDianyinSbtn = (AppCompatRadioButton) this.findViewById(R.id.dianyin_sbtn);
        mKonglingSbtn = (AppCompatRadioButton) this.findViewById(R.id.kongling_sbtn);

        int marginLeft = getMarginLeft();
        marginLeft = marginLeft / 6;

        if (!isShowACC) {
            mAccVoice.setVisibility(GONE);
            mMusicVoiceSeekbar.setVisibility(GONE);
        }

        setMarginLeft(mKtvSbtn, marginLeft);
        setMarginLeft(mRockSbtn, marginLeft);
        setMarginLeft(mDianyinSbtn, marginLeft);
        setMarginLeft(mKonglingSbtn, marginLeft);

//        mPeopleVoiceSeekbar.getThumb().setColorFilter(Color.parseColor("#C7C7C7"), PorterDuff.Mode.SRC_ATOP);
//        mMusicVoiceSeekbar.getThumb().setColorFilter(Color.parseColor("#C7C7C7"), PorterDuff.Mode.SRC_ATOP);

        mPeopleVoiceSeekbar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
//                ZqEngineKit.getInstance().adjustPlaybackSignalVolume(progress);
                mAfterPeopleVoice = progress;
                ZqEngineKit.getInstance().adjustRecordingSignalVolume(progress);
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {

            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {

            }
        });

        mMusicVoiceSeekbar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                mAfterMusicVoice = progress;
                ZqEngineKit.getInstance().adjustAudioMixingPlayoutVolume(progress);
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {

            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {

            }
        });
        mScenesBtnGroup = (RadioGroup) this.findViewById(R.id.scenes_btn_group);

        mScenesBtnGroup.setOnCheckedChangeListener(new RadioGroup.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(RadioGroup group, int checkedId) {
                MyLog.d(TAG, "onCheckedChanged" + " group=" + group + " checkedId=" + checkedId);
                if (checkedId == R.id.default_sbtn) {
                    mAfterMode = Params.AudioEffect.none;
                    ZqEngineKit.getInstance().setAudioEffectStyle(Params.AudioEffect.none);
                } else if (checkedId == R.id.ktv_sbtn) {
                    mAfterMode = Params.AudioEffect.ktv;
                    ZqEngineKit.getInstance().setAudioEffectStyle(Params.AudioEffect.ktv);
                } else if (checkedId == R.id.rock_sbtn) {
                    mAfterMode = Params.AudioEffect.rock;
                    ZqEngineKit.getInstance().setAudioEffectStyle(Params.AudioEffect.rock);
                } else if (checkedId == R.id.dianyin_sbtn) {
                    mAfterMode = Params.AudioEffect.dianyin;
                    ZqEngineKit.getInstance().setAudioEffectStyle(Params.AudioEffect.dianyin);
                } else if (checkedId == R.id.kongling_sbtn) {
                    mAfterMode = Params.AudioEffect.kongling;
                    ZqEngineKit.getInstance().setAudioEffectStyle(Params.AudioEffect.kongling);
                }
            }
        });
    }

    private void setMarginLeft(AppCompatRadioButton view, int marginLeft) {
        RadioGroup.LayoutParams layoutParams = (RadioGroup.LayoutParams) view.getLayoutParams();
        layoutParams.setMargins(marginLeft, 0, 0, 0);
    }

    public void bindData() {
        Params.AudioEffect styleEnum = null;
        if (ZqEngineKit.getInstance().getParams() != null) {
            styleEnum = ZqEngineKit.getInstance().getParams().getStyleEnum();
        }
        if (styleEnum == Params.AudioEffect.dianyin) {
            mScenesBtnGroup.check(R.id.dianyin_sbtn);
        } else if (styleEnum == Params.AudioEffect.kongling) {
            mScenesBtnGroup.check(R.id.kongling_sbtn);
        } else if (styleEnum == Params.AudioEffect.ktv) {
            mScenesBtnGroup.check(R.id.ktv_sbtn);
        } else if (styleEnum == Params.AudioEffect.rock) {
            mScenesBtnGroup.check(R.id.rock_sbtn);
        } else {
            mScenesBtnGroup.check(R.id.default_sbtn);
        }
        mPeopleVoiceSeekbar.setProgress(ZqEngineKit.getInstance().getParams().getRecordingSignalVolume());
        mMusicVoiceSeekbar.setProgress(ZqEngineKit.getInstance().getParams().getAudioMixingPlayoutVolume());

        mBeforeMode = styleEnum;
        mBeforePeopleVoice = ZqEngineKit.getInstance().getParams().getRecordingSignalVolume();
        mBeforeMusicVoice = ZqEngineKit.getInstance().getParams().getAudioMixingPlayoutVolume();

        mAfterMode = styleEnum;
        mAfterPeopleVoice = ZqEngineKit.getInstance().getParams().getRecordingSignalVolume();
        mAfterMusicVoice = ZqEngineKit.getInstance().getParams().getAudioMixingPlayoutVolume();
    }

    public boolean isChange() {
        if (mBeforeMode == mAfterMode && mBeforeMusicVoice == mAfterMusicVoice && mBeforePeopleVoice == mAfterPeopleVoice) {
            return false;
        }
        return true;
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        // 停止播放混音
//        ZqEngineKit.getInstance().pauseAudioMixing();
        EventBus.getDefault().unregister(this);
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        if (!EventBus.getDefault().isRegistered(this)) {
            EventBus.getDefault().register(this);
        }
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onEvent(EngineEvent event) {
        if (EngineEvent.TYPE_USER_SELF_JOIN_SUCCESS == event.getType()) {
            bindData();
        }
    }
}

File: Component/BusiLib/src/main/java-voice/com/component/voice/control/VoiceControlPanelView.kt
package com.component.voice.control

import android.content.Context
import android.support.v7.widget.AppCompatRadioButton
import android.util.AttributeSet
import android.view.View
import android.widget.*

import com.common.utils.U
import com.common.view.ex.ExTextView
import com.component.busilib.R
import com.engine.EngineEvent
import com.engine.Params
import com.kyleduo.switchbutton.SwitchButton
import com.zq.mediaengine.kit.ZqEngineKit

import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode

open abstract class VoiceControlPanelView : ScrollView {

    protected var mEarSb: SwitchButton ?= null
    protected var mEarTv:TextView? = null
    protected var mPeopleVoice: ExTextView? =null
    protected var mPeopleVoiceSeekbar: SeekBar? =null
    protected var mAccVoice: ExTextView? =null
    protected var mMusicVoiceSeekbar: SeekBar? =null

    protected var mScenesBtnGroup: RadioGroup? =null
    protected var mDefaultSbtn: AppCompatRadioButton? =null
    protected var mKtvSbtn: AppCompatRadioButton?=null
    protected var mRockSbtn: AppCompatRadioButton?=null
    protected var mDianyinSbtn: AppCompatRadioButton?=null
    protected var mKonglingSbtn: AppCompatRadioButton?=null

    // 记录值用来标记是否改变
    internal var mBeforeMode: Params.AudioEffect? = null
    internal var mBeforePeopleVoice: Int = 0
    internal var mBeforeMusicVoice: Int = 0

    internal var mAfterMode: Params.AudioEffect? = null
    internal var mAfterPeopleVoice: Int = 0
    internal var mAfterMusicVoice: Int = 0

    internal var isShowACC = true  //是否显示伴奏


    protected open abstract fun getLayout():Int

    protected open fun getMarginLeft():Int{
        return U.getDisplayUtils().screenWidth - U.getDisplayUtils().dip2px((30 + 24).toFloat()) - U.getDisplayUtils().dip2px((53 * 5).toFloat())
    }

    val isChange: Boolean
        get() = !(mBeforeMode == mAfterMode && mBeforeMusicVoice == mAfterMusicVoice && mBeforePeopleVoice == mAfterPeopleVoice)

    constructor(context: Context?) : super(context) {
        init(context)
    }

    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs) {
        val typedArray = context?.obtainStyledAttributes(attrs, R.styleable.VoiceControlPanelView)
        isShowACC = typedArray!!.getBoolean(R.styleable.VoiceControlPanelView_isShowACC, true)
        typedArray?.recycle()
        init(context)
    }

    open fun init(context: Context?) {
        View.inflate(getContext(), getLayout(), this)
        mPeopleVoice = this.findViewById(R.id.people_voice)
        mPeopleVoiceSeekbar = this.findViewById(R.id.people_voice_seekbar)
        mAccVoice = this.findViewById(R.id.acc_voice)
        mMusicVoiceSeekbar = this.findViewById(R.id.music_voice_seekbar)

        mDefaultSbtn = this.findViewById(R.id.default_sbtn)
        mKtvSbtn = this.findViewById(R.id.ktv_sbtn)
        mRockSbtn = this.findViewById(R.id.rock_sbtn)
        mDianyinSbtn = this.findViewById(R.id.liuxing_sbtn)
        mKonglingSbtn = this.findViewById(R.id.kongling_sbtn)

        mEarSb = this.findViewById(R.id.ear_sb)
        mEarTv = this.findViewById(R.id.ear_tv)

        var marginLeft = getMarginLeft()
        marginLeft = marginLeft / 6

        if (!isShowACC) {
            mAccVoice?.visibility = View.GONE
            mMusicVoiceSeekbar?.visibility = View.GONE
        }

        setMarginLeft(mKtvSbtn, marginLeft)
        setMarginLeft(mRockSbtn, marginLeft)
        setMarginLeft(mDianyinSbtn, marginLeft)
        setMarginLeft(mKonglingSbtn, marginLeft)

        mScenesBtnGroup = this.findViewById(R.id.scenes_btn_group)
        setListener()
    }

    protected open fun setListener() {
        mPeopleVoiceSeekbar?.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                //                ZqEngineKit.getInstance().adjustPlaybackSignalVolume(progress);
                mAfterPeopleVoice = progress
                ZqEngineKit.getInstance().adjustRecordingSignalVolume(progress)
//                ZqEngineKit.getInstance().adjustPlaybackSignalVolume(progress)
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {

            }

            override fun onStopTrackingTouch(seekBar: SeekBar) {

            }
        })

        mMusicVoiceSeekbar?.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                mAfterMusicVoice = progress
                ZqEngineKit.getInstance().adjustAudioMixingPlayoutVolume(progress)
                ZqEngineKit.getInstance().adjustAudioMixingPublishVolume(progress,true)
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {

            }

            override fun onStopTrackingTouch(seekBar: SeekBar) {

            }
        })

        mScenesBtnGroup?.setOnCheckedChangeListener { group, checkedId ->
            when (checkedId) {
                R.id.default_sbtn -> {
                    mAfterMode = Params.AudioEffect.none
                    ZqEngineKit.getInstance().setAudioEffectStyle(Params.AudioEffect.none)
                }
                R.id.ktv_sbtn -> {
                    mAfterMode = Params.AudioEffect.ktv
                    ZqEngineKit.getInstance().setAudioEffectStyle(Params.AudioEffect.ktv)
                }
                R.id.rock_sbtn -> {
                    mAfterMode = Params.AudioEffect.rock
                    ZqEngineKit.getInstance().setAudioEffectStyle(Params.AudioEffect.rock)
                }
                R.id.liuxing_sbtn -> {
                    mAfterMode = Params.AudioEffect.liuxing
                    ZqEngineKit.getInstance().setAudioEffectStyle(Params.AudioEffect.liuxing)
                }
                R.id.kongling_sbtn -> {
                    mAfterMode = Params.AudioEffect.kongling
                    ZqEngineKit.getInstance().setAudioEffectStyle(Params.AudioEffect.kongling)
                }
            }
        }

//        mLowLatencySb?.setOnCheckedChangeListener { buttonView, isChecked ->
//            ZqEngineKit.getInstance().setEnableAudioLowLatency(isChecked)
//        }
        mEarSb?.setOnCheckedChangeListener { buttonView, isChecked ->
            // TODO: 测试用途
            ZqEngineKit.getInstance().enableInEarMonitoring(isChecked)
//            if (isChecked) {
//                mMixSb?.setCheckedNoEvent(false)
//            }
//            ZqEngineKit.getInstance().setEnableAudioPreviewLatencyTest(isChecked)
        }
//        mMixSb?.setOnCheckedChangeListener { buttonView, isChecked ->
//            if (isChecked) {
//                mEarSb?.setCheckedNoEvent(false)
//            }
//            ZqEngineKit.getInstance().setEnableAudioMixLatencyTest(isChecked)
//        }
    }

    private fun setMarginLeft(view: AppCompatRadioButton?, marginLeft: Int) {
        val layoutParams = view?.layoutParams as RadioGroup.LayoutParams?
        layoutParams?.setMargins(marginLeft, 0, 0, 0)
    }

    open fun bindData() {
        var styleEnum: Params.AudioEffect? = null
        if (ZqEngineKit.getInstance().params != null) {
            styleEnum = ZqEngineKit.getInstance().params.styleEnum
        }
        if (styleEnum == Params.AudioEffect.liuxing) {
            mScenesBtnGroup?.check(R.id.liuxing_sbtn)
        } else if (styleEnum == Params.AudioEffect.kongling) {
            mScenesBtnGroup?.check(R.id.kongling_sbtn)
        } else if (styleEnum == Params.AudioEffect.ktv) {
            mScenesBtnGroup?.check(R.id.ktv_sbtn)
        } else if (styleEnum == Params.AudioEffect.rock) {
            mScenesBtnGroup?.check(R.id.rock_sbtn)
        } else {
            mScenesBtnGroup?.check(R.id.default_sbtn)
        }
        mPeopleVoiceSeekbar?.progress = ZqEngineKit.getInstance().params.recordingSignalVolume
        mMusicVoiceSeekbar?.progress = ZqEngineKit.getInstance().params.audioMixingPlayoutVolume

        mBeforeMode = styleEnum
        mBeforePeopleVoice = ZqEngineKit.getInstance().params.recordingSignalVolume
        mBeforeMusicVoice = ZqEngineKit.getInstance().params.audioMixingPlayoutVolume

        mAfterMode = styleEnum
        mAfterPeopleVoice = ZqEngineKit.getInstance().params.recordingSignalVolume
        mAfterMusicVoice = ZqEngineKit.getInstance().params.audioMixingPlayoutVolume

        if(ZqEngineKit.getInstance().params.isUseExternalAudio){
            mEarTv?.visibility = View.VISIBLE
            mEarSb?.visibility = View.VISIBLE
            mEarSb?.setCheckedNoEvent(ZqEngineKit.getInstance().params.isEnableInEarMonitoring)
        }else{
            mEarTv?.visibility = View.GONE
            mEarSb?.visibility = View.GONE
        }
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        // 停止播放混音
        //        ZqEngineKit.getInstance().pauseAudioMixing();
        EventBus.getDefault().unregister(this)
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        if (!EventBus.getDefault().isRegistered(this)) {
            EventBus.getDefault().register(this)
        }
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onEvent(event: EngineEvent) {
        if (EngineEvent.TYPE_USER_SELF_JOIN_SUCCESS == event.getType()) {
            bindData()
        }
    }
}
